File: java-workshop/part-1-powermock/src/main/java/service/StockService.java
package service;

import domain.OrderItem;
import domain.Stock;
import domain.StockReservation;
import infrastructure.Database;

import java.util.ArrayList;
import java.util.List;

/**
 * Stock service with hardcoded Database dependency (static calls).
 * Forces PowerMock-style mocking.
 */
public class StockService {

    public List<StockReservation> reserveStock(List<OrderItem> items) {
        List<StockReservation> reservations = new ArrayList<>();

        for (OrderItem item : items) {
            Stock stock = Database.getStock(item.productId());
            if (stock != null && stock.availableQuantity() >= item.quantity()) {
                Database.updateStock(item.productId(), stock.availableQuantity() - item.quantity());
                reservations.add(new StockReservation(item.productId(), item.quantity(), true));
            } else {
                reservations.add(new StockReservation(item.productId(), item.quantity(), false));
            }
        }

        return reservations;
    }

    public void releaseStock(List<StockReservation> reservations) {
        for (StockReservation reservation : reservations) {
            if (!reservation.reserved()) {
                continue;
            }

            Stock stock = Database.getStock(reservation.productId());
            if (stock != null) {
                Database.updateStock(
                    reservation.productId(),
                    stock.availableQuantity() + reservation.quantity()
                );
            }
        }
    }
}

File: csharp-workshop/Part1.StaticDeps/Service/StockService.cs
using Part1.StaticDeps.Domain;
using Part1.StaticDeps.Infrastructure;

namespace Part1.StaticDeps.Service;

/// <summary>
/// Stock service with hardcoded Database dependency (static calls).
/// Requires static mocking (JustMock/Typemock) to test in isolation.
/// Java equivalent requires PowerMock.mockStatic().
/// </summary>
public class StockService
{
    public List<StockReservation> ReserveStock(IReadOnlyList<OrderItem> items)
    {
        var reservations = new List<StockReservation>();

        foreach (var item in items)
        {
            var stock = Database.GetStock(item.ProductId);
            if (stock is not null && stock.AvailableQuantity >= item.Quantity)
            {
                Database.UpdateStock(item.ProductId, stock.AvailableQuantity - item.Quantity);
                reservations.Add(new StockReservation(item.ProductId, item.Quantity, true));
            }
            else
            {
                reservations.Add(new StockReservation(item.ProductId, item.Quantity, false));
            }
        }

        return reservations;
    }

    public void ReleaseStock(IReadOnlyList<StockReservation> reservations)
    {
        foreach (var reservation in reservations)
        {
            if (!reservation.Reserved)
            {
                continue;
            }

            var stock = Database.GetStock(reservation.ProductId);
            if (stock is not null)
            {
                Database.UpdateStock(
                    reservation.ProductId,
                    stock.AvailableQuantity + reservation.Quantity);
            }
        }
    }
}
